src/backend/joanie/core/api.py

```diff
@@ -127,6 +137,124 @@ def perform_create(self, serializer):
+
+        # Once order has been created, if product is not free, create a payment
+        if product.price.amount > 0:
+            order = serializer.instance
+            payment_backend = get_payment_backend()
+
+            if credit_card_id:
+                try:
+                    credit_card = CreditCard.objects.get(
+                        owner=order.owner, uid=credit_card_id
+                    )
+                    payment_info = payment_backend.create_one_click_payment(
+                        request=request,
+                        order=order,
+                        billing_address=billing_address,
+                        credit_card_token=credit_card.token,
+                    )
+                    return Response(
+                        {**serializer.data, "payment_info": payment_info}, status=201
+                    )
+                except (CreditCard.DoesNotExist, NotImplementedError):
+                    pass
+            else:
+                payment_info = payment_backend.create_payment(
+                    request=request, order=order, billing_address=billing_address
+                )
+
+            # Return the fresh new order with payment_info
+            return Response(
+                {**serializer.data, "payment_info": payment_info}, status=201
+            )
+
+        # Else return the fresh new order
+        return Response(serializer.data, status=201)
+
+    @action(detail=True, methods=["POST"])
+    def abort(self, request, uid=None):  # pylint: disable=no-self-use
+        """Abort a pending order and the related payment if there is one."""
+        payment_id = request.data.get("payment_id")
+
+        try:
+            user = models.User.objects.get(username=request.user.username)
+            order = models.Order.objects.get(uid=uid, owner=user)
+        except ObjectDoesNotExist:
+            return Response("Cannot found the order ressource.", status=404)
+
+        if order.state != ORDER_STATE_PENDING:
+            return Response("Cannot abort a not pending order.", status=403)
+
+        if payment_id:
+            payment_backend = get_payment_backend()
+            payment_backend.abort_payment(payment_id)
+
+        order.cancel()
+
+        return Response(status=204)
+
+    @action(detail=True, methods=["GET"])
+    def invoice(self, request, uid=None):  # pylint: disable=no-self-use
+        """
+        Retrieve an invoice through its reference if it is related to the order instance
+        and owned by the authenticated user.
+        """
+        invoice_reference = request.query_params.get("reference")
+
+        if invoice_reference is None:
+            return Response({"reference": "This parameter is required."}, status=400)
+
+        try:
+            user = models.User.objects.get(username=request.user.username)
+            invoice = Invoice.objects.get(
+                reference=invoice_reference, order__uid=uid, order__owner=user
+            )
```

Comment:
You can do it in one query:
```suggestion
            invoice = Invoice.objects.get(
                reference=invoice_reference, order__uid=uid, order__owner__username=request.user.username
            )
```